src/memo/memo.ml

```diff
@@ -245,71 +276,117 @@ module M = struct
-      | Init
-      | Running : Running_state.t * ('a, 'b, 'f) Completion.t -> ('a, 'b, 'f) t
-      | Done of 'b Cached_value.t
+      (* [Considering] marks computations currently being considered (either
+         validated, or run). Note, however, that a stale value of [Considering]
+         may remain if evaluation is cancelled. Only [Considering] constructors
+         with the value of [run] equal to [Run.current ()] should be taken into
+         account. *)
+      | Not_considering
+      | Considering of
+          { run : Run.t
+          ; running : Running_state.t
+          ; completion : ('a, 'b, 'f) Completion.t
+          }
   end =
     State
 
   and Dep_node : sig
     type ('a, 'b, 'f) t =
       { without_state : ('a, 'b, 'f) Dep_node_without_state.t
       ; mutable state : ('a, 'b, 'f) State.t
+      ; mutable last_cached_value : 'b Cached_value.t option
       }
 
     type packed = T : (_, _, _) t -> packed [@@unboxed]
   end =
     Dep_node
 
-  (* We store the last value ['b] we depended on to support early cutoff. *)
+  (* We store the last value ['b Cache_value.t] we depended on to support early
+     cutoff. *)
   and Last_dep : sig
-    type t = T : ('a, 'b, 'f) Dep_node.t * 'b Value.t -> t
+    type t = T : ('a, 'b, 'f) Dep_node.t * 'b Cached_value.t -> t
   end =
     Last_dep
 end
 
 module State = M.State
 module Running_state = M.Running_state
 module Dep_node = M.Dep_node
-module Deps_so_far = M.Deps_so_far
 module Last_dep = M.Last_dep
+module Deps_so_far = M.Deps_so_far
 
 let no_deps_so_far : Deps_so_far.t = { set = Id.Set.empty; deps_reversed = [] }
 
+let currently_considering (v : _ State.t) : _ State.t =
+  match v with
+  | Not_considering -> Not_considering
+  | Considering { run; _ } as running ->
+    if Run.is_current run then
+      running
+    else
+      Not_considering
+
+let get_cached_value_in_current_cycle (dep_node : _ Dep_node.t) =
+  match dep_node.last_cached_value with
+  | None -> None
+  | Some cv ->
+    if Run.is_current cv.last_validated_at then
+      Some cv
+    else
+      None
+
 module Cached_value = struct
   include M.Cached_value
 
-  let create x ~deps = { deps; value = x; calculated_at = Run.current () }
+  let capture_dep_values ~deps_rev =
+    List.rev_map deps_rev ~f:(function Dep_node.T dep_node ->
+        ( match get_cached_value_in_current_cycle dep_node with
+        | None ->
+          Code_error.raise
+            "Attempted to create a cached value based on some stale inputs" []
+        | Some cv -> Last_dep.T (dep_node, cv) ))
+
+  let create x ~deps_rev =
+    { deps = capture_dep_values ~deps_rev
+    ; value = x
+    ; last_validated_at = Run.current ()
+    ; id = Value_id.create ()
+    }
+
+  let confirm_old_value t ~deps_rev =
+    { value = t.value
+    ; id = t.id
+    ; last_validated_at = Run.current ()
+    ; deps = capture_dep_values ~deps_rev
+    }
 
   let dep_changed (type a) (node : (_, a, _) Dep_node.t) prev_output curr_output
```

Comment:
Probably should be renamed: we don't look at `node` as a dependency when we call this function.

Reply:
Renamed to `value_changed`.